File: corelib/hash.rb
class Hash
  include Enumerable

  %x{
    var $hash = Opal.hash = function() {
      if (arguments.length == 1 && arguments[0]._klass == Hash) {
        return arguments[0];
      }

      var hash   = new Hash._alloc,
          keys   = [],
          assocs = {};

      hash.map   = assocs;
      hash.keys  = keys;

      if (arguments.length == 1 && arguments[0]._isArray) {
        var args = arguments[0];

        for (var i = 0, length = args.length; i < length; i++) {
          var key = args[i][0], obj = args[i][1];

          if (assocs[key] == null) {
            keys.push(key);
          }

          assocs[key] = obj;
        }
      }
      else {
        for (var i = 0, length = arguments.length, key; i < length; i++) {
          var key = arguments[i], obj = arguments[++i];

          if (assocs[key] == null) {
            keys.push(key);
          }

          assocs[key] = obj;
        }
      }

      return hash;
    };
  }

  # hash2 is a faster creator for hashes that just use symbols and
  # strings as keys. The map and keys array can be constructed at
  # compile time, so they are just added here by the constructor
  # function
  %x{
    var $hash2 = Opal.hash2 = function(keys, map) {
      var hash = new Hash._alloc;
      hash.keys = keys;
      hash.map = map;
      return hash;
    };
  }

  `var $hasOwn = {}.hasOwnProperty`

  def self.[](*objs)
    `$hash.apply(null, objs)`
  end

  def self.allocate
    %x{
      var hash = new #{self}._alloc;
      hash.map = {};
      hash.keys = [];
      return hash;
    }
  end

  def initialize(defaults = undefined, &block)
    %x{
      if (defaults != null) {
        if (defaults.constructor == Object) {
          var map = #{self}.map, keys = #{self}.keys;

          for (var key in defaults) {
            keys.push(key);
            map[key] = defaults[key];
          }
        }
        else {
          #{self}.none = defaults;
        }
      }
      else if (block !== nil) {
          #{self}.proc = block;
      }

      return #{self};
    }
  end

  def ==(other)
    %x{
      if (#{self} === other) {
        return true;
      }

      if (!other.map || !other.keys) {
        return false;
      }

      if (#{self}.keys.length !== other.keys.length) {
        return false;
      }

      var map  = #{self}.map,
          map2 = other.map;

      for (var i = 0, length = #{self}.keys.length; i < length; i++) {
        var key = #{self}.keys[i], obj = map[key], obj2 = map2[key];

        if (#{`obj` != `obj2`}) {
          return false;
        }
      }

      return true;
    }
  end

  def [](key)
    %x{
      var map = #{self}.map;

      if ($hasOwn.call(map, key)) {
        return map[key];
      }

      var proc = #{@proc};

      if (proc !== nil) {
        return #{ `proc`.call self, key };
      }

      return #{@none};
    }
  end

  def []=(key, value)
    %x{
      var map = #{self}.map;

      if (!$hasOwn.call(map, key)) {
        #{self}.keys.push(key);
      }

      map[key] = value;

      return value;
    }
  end

  def assoc(object)
    %x{
      var keys = #{self}.keys, key;

      for (var i = 0, length = keys.length; i < length; i++) {
        key = keys[i];

        if (#{`key` == object}) {
          return [key, #{self}.map[key]];
        }
      }

      return nil;
    }
  end

  def clear
    %x{
      #{self}.map = {};
      #{self}.keys = [];
      return #{self};
    }
  end

  def clone
    %x{
      var result = new self._klass._alloc();

      result.map = {}; result.keys = [];

      var map    = #{self}.map,
          map2   = result.map,
          keys2  = result.keys;

      for (var i = 0, length = #{self}.keys.length; i < length; i++) {
        keys2.push(#{self}.keys[i]);
        map2[#{self}.keys[i]] = map[#{self}.keys[i]];
      }

      return result;
    }
  end

  def default(val = undefined)
    @none
  end

  def default=(object)
    @none = object
  end

  def default_proc
    @proc
  end

  def default_proc=(proc)
    @proc = proc
  end

  def delete(key)
    %x{
      var map  = #{self}.map, result = map[key];

      if (result != null) {
        delete map[key];
        #{self}.keys.$delete(key);

        return result;
      }

      return nil;
    }
  end

  def delete_if(&block)
    return enum_for :delete_if unless block

    %x{
      var map = #{self}.map, keys = #{self}.keys, value;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key];

        if ((value = block(key, obj)) === $breaker) {
          return $breaker.$v;
        }

        if (value !== false && value !== nil) {
          keys.splice(i, 1);
          delete map[key];

          length--;
          i--;
        }
      }

      return #{self};
    }
  end

  alias dup clone

  def each(&block)
    return enum_for :each unless block

    %x{
      var map = #{self}.map, keys = #{self}.keys;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i];

        #{yield [`key`, `map[key]`]};
      }

      return #{self};
    }
  end

  def each_key(&block)
    return enum_for :each_key unless block

    %x{
      var keys = #{self}.keys;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i];

        if (block(key) === $breaker) {
          return $breaker.$v;
        }
      }

      return #{self};
    }
  end

  alias each_pair each

  def each_value(&block)
    return enum_for :each_value unless block

    %x{
      var map = #{self}.map, keys = #{self}.keys;

      for (var i = 0, length = keys.length; i < length; i++) {
        if (block(map[keys[i]]) === $breaker) {
          return $breaker.$v;
        }
      }

      return #{self};
    }
  end

  def empty?
    `#{self}.keys.length === 0`
  end

  alias eql? ==

  def fetch(key, defaults = undefined, &block)
    %x{
      var value = #{self}.map[key];

      if (value != null) {
        return value;
      }

      if (block !== nil) {
        var value;

        if ((value = block(key)) === $breaker) {
          return $breaker.$v;
        }

        return value;
      }

      if (defaults != null) {
        return defaults;
      }

      #{ raise KeyError, "key not found" };
    }
  end

  def flatten(level=undefined)
    %x{
      var map = #{self}.map, keys = #{self}.keys, result = [];

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], value = map[key];

        result.push(key);

        if (value._isArray) {
          if (level == null || level === 1) {
            result.push(value);
          }
          else {
            result = result.concat(#{`value`.flatten(`level - 1`)});
          }
        }
        else {
          result.push(value);
        }
      }

      return result;
    }
  end

  def has_key?(key)
    `$hasOwn.call(#{self}.map, key)`
  end

  def has_value?(value)
    %x{
      for (var assoc in #{self}.map) {
        if (#{`#{self}.map[assoc]` == value}) {
          return true;
        }
      }

      return false;
    }
  end

  def hash
    `#{self}._id`
  end

  alias include? has_key?

  def index(object)
    %x{
      var map = #{self}.map, keys = #{self}.keys;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i];

        if (#{object == `map[key]`}) {
          return key;
        }
      }

      return nil;
    }
  end

  def indexes(*keys)
    %x{
      var result = [], map = #{self}.map, val;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], val = map[key];

        if (val != null) {
          result.push(val);
        }
        else {
          result.push(#{self}.none);
        }
      }

      return result;
    }
  end

  alias indices indexes

  def inspect
    %x{
      var inspect = [], keys = #{self}.keys, map = #{self}.map;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], val = map[key];

        if (val === #{self}) {
          inspect.push(#{`key`.inspect} + '=>' + '{...}');
        } else {
          inspect.push(#{`key`.inspect} + '=>' + #{`map[key]`.inspect});
        }
      }

      return '{' + inspect.join(', ') + '}';
    }
  end

  def invert
    %x{
      var result = $hash(), keys = #{self}.keys, map = #{self}.map,
          keys2 = result.keys, map2 = result.map;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key];

        keys2.push(obj);
        map2[obj] = key;
      }

      return result;
    }
  end

  def keep_if(&block)
    return enum_for :keep_if unless block

    %x{
      var map = #{self}.map, keys = #{self}.keys, value;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key];

        if ((value = block(key, obj)) === $breaker) {
          return $breaker.$v;
        }

        if (value === false || value === nil) {
          keys.splice(i, 1);
          delete map[key];

          length--;
          i--;
        }
      }

      return #{self};
    }
  end

  alias key index

  alias key? has_key?

  def keys
    %x{
      return #{self}.keys.slice(0);
    }
  end

  def length
    %x{
      return #{self}.keys.length;
    }
  end

  alias member? has_key?

  def merge(other, &block)
    %x{
      var keys = #{self}.keys, map = #{self}.map,
          result = $hash(), keys2 = result.keys, map2 = result.map;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i];

        keys2.push(key);
        map2[key] = map[key];
      }

      var keys = other.keys, map = other.map;

      if (block === nil) {
        for (var i = 0, length = keys.length; i < length; i++) {
          var key = keys[i];

          if (map2[key] == null) {
            keys2.push(key);
          }

          map2[key] = map[key];
        }
      }
      else {
        for (var i = 0, length = keys.length; i < length; i++) {
          var key = keys[i];

          if (map2[key] == null) {
            keys2.push(key);
            map2[key] = map[key];
          }
          else {
            map2[key] = block(key, map2[key], map[key]);
          }
        }
      }

      return result;
    }
  end

  def merge!(other, &block)
    %x{
      var keys = #{self}.keys, map = #{self}.map,
          keys2 = other.keys, map2 = other.map;

      if (block === nil) {
        for (var i = 0, length = keys2.length; i < length; i++) {
          var key = keys2[i];

          if (map[key] == null) {
            keys.push(key);
          }

          map[key] = map2[key];
        }
      }
      else {
        for (var i = 0, length = keys2.length; i < length; i++) {
          var key = keys2[i];

          if (map[key] == null) {
            keys.push(key);
            map[key] = map2[key];
          }
          else {
            map[key] = block(key, map[key], map2[key]);
          }
        }
      }

      return #{self};
    }
  end

  def rassoc(object)
    %x{
      var keys = #{self}.keys, map = #{self}.map;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key];

        if (#{`obj` == object}) {
          return [key, obj];
        }
      }

      return nil;
    }
  end

  def reject(&block)
    return enum_for :reject unless block

    %x{
      var keys = #{self}.keys, map = #{self}.map,
          result = $hash(), map2 = result.map, keys2 = result.keys;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key], value;

        if ((value = block(key, obj)) === $breaker) {
          return $breaker.$v;
        }

        if (value === false || value === nil) {
          keys2.push(key);
          map2[key] = obj;
        }
      }

      return result;
    }
  end

  def replace(other)
    %x{
      var map = #{self}.map = {}, keys = #{self}.keys = [];

      for (var i = 0, length = other.keys.length; i < length; i++) {
        var key = other.keys[i];
        keys.push(key);
        map[key] = other.map[key];
      }

      return #{self};
    }
  end

  def select(&block)
    return enum_for :select unless block

    %x{
      var keys = #{self}.keys, map = #{self}.map,
          result = $hash(), map2 = result.map, keys2 = result.keys;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key], value;

        if ((value = block(key, obj)) === $breaker) {
          return $breaker.$v;
        }

        if (value !== false && value !== nil) {
          keys2.push(key);
          map2[key] = obj;
        }
      }

      return result;
    }
  end

  def select!(&block)
    return enum_for :select! unless block

    %x{
      var map = #{self}.map, keys = #{self}.keys, value, result = nil;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key];

        if ((value = block(key, obj)) === $breaker) {
          return $breaker.$v;
        }

        if (value === false || value === nil) {
          keys.splice(i, 1);
          delete map[key];

          length--;
          i--;
          result = #{self}
        }
      }

      return result;
    }
  end

  def shift
    %x{
      var keys = #{self}.keys, map = #{self}.map;

      if (keys.length) {
        var key = keys[0], obj = map[key];

        delete map[key];
        keys.splice(0, 1);

        return [key, obj];
      }

      return nil;
    }
  end

  alias size length

  def to_a
    %x{
      var keys = #{self}.keys, map = #{self}.map, result = [];

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i];
        result.push([key, map[key]]);
      }

      return result;
    }
  end

  def to_hash
    self
  end

  def to_n
    %x{
      var result = {}, keys = #{self}.keys, map = #{self}.map, bucket, value;

      for (var i = 0, length = keys.length; i < length; i++) {
        var key = keys[i], obj = map[key];

        if (obj.$to_n) {
          result[key] = #{`obj`.to_n};
        }
        else {
          result[key] = obj;
        }
      }

      return result;
    }
  end

  alias to_s inspect

  alias update merge!

  def value?(value)
    %x{
      var map = #{self}.map;

      for (var assoc in map) {
        var v = map[assoc];
        if (#{`v` == value}) {
          return true;
        }
      }

      return false;
    }
  end

  alias values_at indexes

  def values
    %x{
      var map    = #{self}.map,
          result = [];

      for (var key in map) {
        result.push(map[key]);
      }

      return result;
    }
  end
end

File: corelib/hash.js
var RubyHash;

// Returns new hash with values passed from ruby
VM.H = function() {
  var hash = new RubyHash.$a(), key, val, args = ArraySlice.call(arguments);
  var assocs = hash.map = {};
  hash.none = Qnil;

  for (var i = 0, ii = args.length; i < ii; i++) {
    key = args[i];
    val = args[i + 1];
    i++;
    assocs[key] = [key, val];
  }

  return hash;
};

function hash_s_create() {
  return VM.H.apply(null, ArraySlice.call(arguments));
}

function hash_alloc() {
  return VM.H();
}

function hash_new() {
  return VM.H();
}

function hash_equal(other) {
  var self = this;
  if (self === other) return true;
  if (!other.map) return false;

  var map = self.map, map2 = other.map;

  for (var assoc in map) {
    if (!map2[assoc]) return false;
    var obj = map[assoc][1], obj2 = map2[assoc][1];
    if (!obj.m$eq$(obj2)) return false;
  }

  return true;
}

function hash_aref(key) {
  var hash = key, val;
  if (val = this.map[hash]) return val[1];
  return this.none;
}

function hash_aset(key, value) {
  var hash = key;
  this.map[hash] = [key, value];
  return value;
}

function hash_assoc(object) {
  for (var assoc in this.map) {
    var bucket = this.map[assoc];
    if (bucket[0].m$eq$(object)) return [bucket[0], bucket[1]];
  }
  return nil;
}

function hash_clear() {
  this.map = {};
  return this;
}

function hash_clone() {
  var result = VM.H(), map = this.map, map2 = result.map;
  for (var assoc in map) map2[assoc] = [map[assoc][0], map[assoc][1]];
  return result;
}

function hash_default() {
  return this.none;
}

function hash_set_default(object) {
  return this.none = object;
}

function hash_default_proc() {
  return this.proc;
}

function hash_set_default_proc(proc) {
  this.proc = proc;
}

function hash_delete(key) {
  var hash = key, bucket, ret;
  if (bucket = this.map[hash]) {
    ret = bucket[1];
    delete this.map[hash];
    return ret;
  }
  return this.none;
}

function hash_delete_if() {
  var self = this, iterator = hash_delete_if.proc, map = self.map, bucket;
  if (!iterator) return self.m$enum_for("delete_if");

  var context = iterator.$s, val;
  hash_delete_if.proc = 0;

  for (var assoc in map) {
    bucket = map[assoc];
    if ((val = iterator.call(context, bucket[0], bucket[1])) === breaker) return breaker.$v;
    if (val !== false && val !== nil) delete map[assoc];
  }
  return self;
}

function hash_each() {
  var self = this, iterator = hash_each.proc, map = self.map, bucket;
  if (!iterator) return self.m$enum_for("each");

  var context = iterator.$S;
  hash_each.proc = 0;

  for (var assoc in map) {
    bucket = map[assoc];
    if (iterator.call(context, bucket[0], bucket[1]) === breaker)
      return breaker.$v;
  }
  return self;
}

function hash_each_key() {
  var self = this, iterator = hash_each_key.proc, map = self.map;
  if (!iterator) return self.m$enum_for("each_key");

  var context = iterator.$S;
  hash_each_key.proc = 0;

  for (var assoc in map) {
    if (iterator.call(context, map[assoc][0]) === breaker) return breaker.$v;
  }
  return self;
}

function hash_each_value() {
  var self = this, iterator = hash_each_value.proc, map = self.map;
  if (!iterator) return self.m$enum_for("each_value");

  var context = iterator.$S;
  hash_each_value.proc = 0;

  for (var assoc in map) {
    if (iterator.call(context, map[assoc][1]) === breaker) return breaker.$v;
  }
  return self;
}

function hash_empty_p() {
  for (var assoc in this.map) return false;
  return true;
}

function hash_fetch(key, defaults) {
  var self = this, map = self.map, bucket = map[key];
  if (bucket) return bucket[1];

  if (hash_fetch.proc) {
    var iterator = hash_fetch.proc, val, context = iterator.$S;
    hash_fetch.proc = 0;

    if ((val = iterator.call(context, key)) === breaker) return breaker.$v;
    return val;
  }

  if (defaults !== undefined) return defaults;
  rb_raise(rb_eKeyError, 'key not found');
}

function hash_flatten(level) {
  var self = this, map = self.map, result = [];
  for (var assoc in map) {
    var bucket = map[assoc], key = bucket[0], value = bucket[1];
    result.push(key);

    if (value.$f & T_ARRAY) {
      if (level === undefined || level === 1) result.push(value);
      else result = result.concat(value.m$flatten(level - 1));
    }
    else {
      result.push(value);
    }
  }
  return result;
}

function hash_has_key(key) {
  return !!this.map[key];
}

function hash_has_value(value) {
  for (var assoc in this.map) {
    if (this.map[assoc][1].m$eq$(value)) return true;
  }
  return false;
}

function hash_hash() {
  return this.$id;
}

function hash_inspect() {
  var self = this, parts = [], map = self.map;
  for (var assoc in map) {
    var bucket = map[assoc];
    parts.push(bucket[0].m$inspect() + '=>' + bucket[1].m$inspect());
  }
  return '{' + parts.join(', ') + '}';
}

function hash_invert() {
  var self = this, map = self.map, result = VM.H(), map2 = result.map;
  for (var assoc in map) {
    var bucket = map[assoc];
    map2[bucket[1]] = [bucket[0], bucket[1]];
  }
  return result;
}

function hash_key(object) {
  for (var assoc in this.map) {
    var bucket = this.map[assoc];
    if (object.m$eq$(bucket[1])) return bucket[0];
  }
  return nil;
}

function hash_keys() {
  var result = [];
  for (var assoc in this.map) result.push(this.map[assoc][0]);
  return result;
}

function hash_length() {
  var length = 0;
  for (var assoc in this.map) length++;
  return length;
}

function hash_merge(other) {
  var self = this, map = self.map, result = VM.H(), result_map = result.map;

  for (var assoc in map) {
    var bucket = map[assoc];
    result_map[assoc] = [bucket[0], bucket[1]];
  }

  map = other.map;

  for (var assoc in map) {
    var bucket = map[assoc];
    result_map[assoc] = [bucket[0], bucket[1]];
  }

  return result;
}

function hash_update(other) {
  var self = this, map = self.map, map2 = other.map;

  for (var assoc in map2) {
    var bucket = map2[assoc]
    map[assoc] = [bucket[0], bucket[1]];
  }
  return self;
}

function hash_rassoc(object) {
  var self = this, map = self.map;
  for (var assoc in map) {
    var bucket = map[assoc];
    if (bucket[1].m$eq$(object)) return [bucket[0], bucket[1]];
  }
  return nil;
}

function hash_replace(other) {
  var self = this, map = self.map = {};
  for (var assoc in other.map) {
    var bucket = other.map[assoc];
    map[assoc] = [bucket[0], bucket[1]];
  }
  return self;
}

function hash_to_a() {
  var self = this, map = self.map, ary = [];
  for (var assoc in map) {
    var bucket = map[assoc];
    ary.push([bucket[0], bucket[1]]);
  }
  return ary;
}

function hash_to_hash() {
  return this;
}

function hash_values() {
  var self = this, map = self.map, values = [];
  for (var assoc in map) values.push(map[assoc][1]);
  return values;
}

function init_hash() {
  RubyHash = define_class(rb_cObject, 'Hash', rb_cObject);
  rb_include_module(RubyHash, RubyEnumerable);

  define_singleton_method(RubyHash, 'm$aref$', hash_s_create);
  define_singleton_method(RubyHash, 'm$allocate', hash_alloc);
  define_singleton_method(RubyHash, 'm$new', hash_new);

  define_methods(RubyHash, {
    'm$eq$': hash_equal,
    'm$aref$': hash_aref,
    'm$aset$': hash_aset,
    'm$assoc': hash_assoc,
    'm$clear': hash_clear,
    'm$clone': hash_clone,
    'm$default': hash_default,
    'm$default$e': hash_set_default,
    'm$default_proc': hash_default_proc,
    'm$default_proc$e': hash_set_default_proc,
    'm$delete': hash_delete,
    'm$delete_if': hash_delete_if,
    'm$each': hash_each,
    'm$each_key': hash_each_key,
    'm$each_pair': hash_each,
    'm$each_value': hash_each_value,
    'm$empty$p': hash_empty_p,
    'm$eql$p': hash_equal,
    'm$fetch': hash_fetch,
    'm$flatten': hash_flatten,
    'm$has_key$p': hash_has_key,
    'm$has_value$p': hash_has_value,
    'm$hash': hash_hash,
    'm$include$p': hash_has_key,
    'm$inspect': hash_inspect,
    'm$invert': hash_invert,
    'm$key': hash_key,
    'm$key$p': hash_has_key,
    'm$keys': hash_keys,
    'm$length': hash_length,
    'm$member$p': hash_has_key,
    'm$merge': hash_merge,
    'm$merge$b': hash_update,
    'm$rassoc': hash_rassoc,
    'm$replace': hash_replace,
    'm$size': hash_length,
    'm$to_a': hash_to_a,
    'm$to_hash': hash_to_hash,
    'm$to_s': hash_inspect,
    'm$update': hash_update,
    'm$values': hash_values
  });
}
